File: apps/ccusage/src/adapter/codebuff/parser.ts
import { stat } from 'node:fs/promises';
import path from 'node:path';
import { readTextFile } from '@ccusage/internal/fs';
import { compareStrings } from '@ccusage/internal/sort';
import { getDefaultWorkerThreadCount, mapWithConcurrency } from '@ccusage/internal/workers';
import { Result } from '@praha/byethrow';
import { createFixture } from 'fs-fixture';
import { discoverCodebuffChatFiles } from './paths.ts';

export type CodebuffUsageEntry = {
	timestamp: string;
	sessionId: string;
	model: string;
	provider: string;
	credits: number;
	inputTokens: number;
	outputTokens: number;
	cacheCreationInputTokens: number;
	cacheReadInputTokens: number;
	dedupKey: string;
};

type AssistantUsage = {
	model?: string;
	credits: number;
	inputTokens: number;
	outputTokens: number;
	cacheCreationInputTokens: number;
	cacheReadInputTokens: number;
};

const DEFAULT_CODEBUFF_MODEL = 'codebuff-unknown';

function asRecord(value: unknown): Record<string, unknown> | undefined {
	if (value == null || typeof value !== 'object' || Array.isArray(value)) {
		return undefined;
	}
	return value as Record<string, unknown>;
}

function readString(record: Record<string, unknown>, key: string): string | undefined {
	const value = record[key];
	return typeof value === 'string' && value !== '' ? value : undefined;
}

function toPositiveNumber(value: unknown): number {
	return typeof value === 'number' && Number.isFinite(value) && value > 0 ? value : 0;
}

function pickPositiveInteger(record: Record<string, unknown>, keys: readonly string[]): number {
	for (const key of keys) {
		const value = toPositiveNumber(record[key]);
		if (value > 0) {
			return Math.trunc(value);
		}
	}
	return 0;
}

function emptyUsage(): AssistantUsage {
	return {
		credits: 0,
		inputTokens: 0,
		outputTokens: 0,
		cacheCreationInputTokens: 0,
		cacheReadInputTokens: 0,
	};
}

function hasSignal(usage: AssistantUsage): boolean {
	return (
		usage.inputTokens > 0 ||
		usage.outputTokens > 0 ||
		usage.cacheCreationInputTokens > 0 ||
		usage.cacheReadInputTokens > 0 ||
		usage.credits > 0
	);
}

function mergeFallback(target: AssistantUsage, fallback: AssistantUsage): void {
	target.inputTokens ||= fallback.inputTokens;
	target.outputTokens ||= fallback.outputTokens;
	target.cacheCreationInputTokens ||= fallback.cacheCreationInputTokens;
	target.cacheReadInputTokens ||= fallback.cacheReadInputTokens;
	target.credits ||= fallback.credits;
	target.model ??= fallback.model;
}

function parseTimestampValue(value: unknown): string | undefined {
	if (typeof value === 'string') {
		const ms = Date.parse(value);
		return Number.isFinite(ms) ? new Date(ms).toISOString() : undefined;
	}
	if (typeof value === 'number' && Number.isFinite(value) && value > 0) {
		const ms = value < 10_000_000_000 ? value * 1000 : value;
		return new Date(ms).toISOString();
	}
	return undefined;
}

export function parseCodebuffChatIdTimestamp(chatId: string): string | undefined {
	const tIndex = chatId.indexOf('T');
	if (tIndex === -1) {
		return undefined;
	}
	const datePart = chatId.slice(0, tIndex);
	const timePart = chatId.slice(tIndex).replace('-', ':').replace('-', ':');
	return parseTimestampValue(`${datePart}${timePart}`);
}

function getMessageTimestamp(message: Record<string, unknown>): string | undefined {
	return (
		parseTimestampValue(message.timestamp) ??
		parseTimestampValue(message.createdAt) ??
		parseTimestampValue(asRecord(message.metadata)?.timestamp)
	);
}

function deriveCodebuffContext(filePath: string): {
	channel: string;
	project: string;
	chatId: string;
	sessionId: string;
} {
	const chatIdBasename = path.basename(path.dirname(filePath));
	const chatId = chatIdBasename === '' ? 'unknown' : chatIdBasename;
	const chatsDir = path.dirname(path.dirname(filePath));
	const projectDir = path.dirname(chatsDir);
	const projectBasename = path.basename(projectDir);
	const project = projectBasename === '' ? 'unknown' : projectBasename;
	const projectsDir = path.dirname(projectDir);
	const channelDir = path.dirname(projectsDir);
	const channelBasename = path.basename(channelDir);
	const channel = channelBasename === '' ? 'manicode' : channelBasename;
	return {
		channel,
		project,
		chatId,
		sessionId: `${channel}/${project}/${chatId}`,
	};
}

function isAssistantMessage(message: Record<string, unknown>): boolean {
	const role = readString(message, 'variant') ?? readString(message, 'role') ?? '';
	return role === 'ai' || role === 'agent' || role === 'assistant';
}

function parseUsageObject(value: unknown): AssistantUsage {
	const usage = emptyUsage();
	const record = asRecord(value);
	if (record == null) {
		return usage;
	}
	usage.inputTokens = pickPositiveInteger(record, [
		'inputTokens',
		'input_tokens',
		'promptTokens',
		'prompt_tokens',
	]);
	usage.outputTokens = pickPositiveInteger(record, [
		'outputTokens',
		'output_tokens',
		'completionTokens',
		'completion_tokens',
	]);
	usage.cacheReadInputTokens =
		pickPositiveInteger(record, ['cacheReadInputTokens', 'cache_read_input_tokens']) ||
		pickPositiveInteger(asRecord(record.promptTokensDetails) ?? {}, ['cachedTokens']) ||
		pickPositiveInteger(asRecord(record.prompt_tokens_details) ?? {}, ['cached_tokens']);
	usage.cacheCreationInputTokens = pickPositiveInteger(record, [
		'cacheCreationInputTokens',
		'cache_creation_input_tokens',
		'cacheCreationTokens',
		'cache_creation_tokens',
		'cachedTokensCreated',
		'cached_tokens_created',
	]);
	usage.credits = toPositiveNumber(record.credits);
	usage.model = readString(record, 'model');
	return usage;
}

function extractUsageFromRunState(metadata: Record<string, unknown>): AssistantUsage | undefined {
	const runState = asRecord(metadata.runState);
	const sessionState = asRecord(runState?.sessionState);
	const mainAgentState = asRecord(sessionState?.mainAgentState);
	const history = mainAgentState?.messageHistory;
	if (!Array.isArray(history)) {
		return undefined;
	}

	const usage = emptyUsage();
	let found = false;
	for (const item of history.toReversed()) {
		const entry = asRecord(item);
		if (entry == null || readString(entry, 'role') !== 'assistant') {
			continue;
		}
		const providerOptions = asRecord(entry.providerOptions);
		if (providerOptions == null) {
			continue;
		}
		const entryUsage = emptyUsage();
		mergeFallback(entryUsage, parseUsageObject(providerOptions.usage));
		const codebuff = asRecord(providerOptions.codebuff);
		mergeFallback(entryUsage, parseUsageObject(codebuff?.usage));
		entryUsage.model = readString(codebuff ?? {}, 'model') ?? entryUsage.model;
		if (hasSignal(entryUsage) || entryUsage.model != null) {
			found = true;
		}
		mergeFallback(usage, entryUsage);
	}
	return found ? usage : undefined;
}

function extractAssistantUsage(message: Record<string, unknown>): AssistantUsage {
	const usage = emptyUsage();
	const metadata = asRecord(message.metadata);
	if (metadata != null) {
		usage.model = readString(metadata, 'model');
		mergeFallback(usage, parseUsageObject(metadata.usage));
		mergeFallback(usage, parseUsageObject(asRecord(metadata.codebuff)?.usage));
		const runStateUsage = extractUsageFromRunState(metadata);
		if (runStateUsage != null) {
			mergeFallback(usage, runStateUsage);
		}
	}
	const credits = toPositiveNumber(message.credits);
	if (credits > 0 && usage.credits <= 0) {
		usage.credits = credits;
	}
	return usage;
}

function inferProvider(model: string): string {
	const normalized = model.toLowerCase();
	if (
		normalized.startsWith('claude-') ||
		normalized.startsWith('anthropic/') ||
		normalized.startsWith('anthropic.')
	) {
		return 'anthropic';
	}
	if (
		normalized.startsWith('gpt-') ||
		normalized.startsWith('o1') ||
		normalized.startsWith('o3') ||
		normalized.startsWith('o4') ||
		normalized.startsWith('openai/')
	) {
		return 'openai';
	}
	if (normalized.startsWith('gemini') || normalized.startsWith('google/')) {
		return 'google';
	}
	if (normalized.startsWith('grok') || normalized.startsWith('xai/')) {
		return 'xai';
	}
	return 'unknown';
}

function getDedupKey(
	message: Record<string, unknown>,
	sessionId: string,
	timestamp: string,
	model: string,
	usage: AssistantUsage,
	ordinal: number,
): string {
	const messageId = readString(message, 'id');
	if (messageId != null) {
		return `codebuff:${sessionId}:${messageId}`;
	}
	return `codebuff:${sessionId}:${timestamp}:${model}:${ordinal}:${usage.inputTokens}:${usage.outputTokens}:${usage.cacheReadInputTokens}:${usage.cacheCreationInputTokens}`;
}

async function getFileModifiedTimestamp(filePath: string): Promise<string> {
	const result = await Result.try({
		try: stat(filePath),
		catch: (error) => error,
	});
	if (Result.isFailure(result)) {
		return new Date(0).toISOString();
	}
	return result.value.mtime.toISOString();
}

async function loadCodebuffChatFile(filePath: string): Promise<CodebuffUsageEntry[]> {
	const readResult = await Result.try({
		try: readTextFile(filePath),
		catch: (error) => error,
	});
	if (Result.isFailure(readResult)) {
		return [];
	}

	const parseResult = Result.try({
		try: () => JSON.parse(readResult.value) as unknown,
		catch: (error) => error,
	})();
	if (Result.isFailure(parseResult) || !Array.isArray(parseResult.value)) {
		return [];
	}

	const context = deriveCodebuffContext(filePath);
	const chatTimestamp = parseCodebuffChatIdTimestamp(context.chatId);
	const fileTimestamp = await getFileModifiedTimestamp(filePath);
	const entries: CodebuffUsageEntry[] = [];
	for (const [ordinal, rawMessage] of parseResult.value.entries()) {
		const message = asRecord(rawMessage);
		if (message == null || !isAssistantMessage(message)) {
			continue;
		}
		const usage = extractAssistantUsage(message);
		if (!hasSignal(usage)) {
			continue;
		}
		const timestamp = getMessageTimestamp(message) ?? chatTimestamp ?? fileTimestamp;
		const model = usage.model ?? DEFAULT_CODEBUFF_MODEL;
		entries.push({
			timestamp,
			sessionId: context.sessionId,
			model,
			provider: inferProvider(model),
			credits: usage.credits,
			inputTokens: usage.inputTokens,
			outputTokens: usage.outputTokens,
			cacheCreationInputTokens: usage.cacheCreationInputTokens,
			cacheReadInputTokens: usage.cacheReadInputTokens,
			dedupKey: getDedupKey(message, context.sessionId, timestamp, model, usage, ordinal),
		});
	}
	return entries;
}

export async function loadCodebuffUsageEntries(): Promise<CodebuffUsageEntry[]> {
	const files = await discoverCodebuffChatFiles();
	const entryGroups = await mapWithConcurrency(
		files,
		getDefaultWorkerThreadCount(files.length),
		loadCodebuffChatFile,
	);
	const deduped = new Map<string, CodebuffUsageEntry>();
	for (const entry of entryGroups.flat()) {
		deduped.set(entry.dedupKey, entry);
	}
	return Array.from(deduped.values()).sort(
		(a, b) => compareStrings(a.timestamp, b.timestamp) || compareStrings(a.dedupKey, b.dedupKey),
	);
}

if (import.meta.vitest != null) {
	describe('loadCodebuffUsageEntries', () => {
		afterEach(() => {
			vi.unstubAllEnvs();
		});

		it('loads assistant usage from chat-messages JSON files', async () => {
			await using fixture = await createFixture({
				projects: {
					'project-a': {
						chats: {
							'2026-01-02T03-04-05.000Z': {
								'chat-messages.json': JSON.stringify([
									{ role: 'user', text: 'hello' },
									{
										id: 'assistant-message',
										role: 'assistant',
										timestamp: '2026-01-02T03:04:06.000Z',
										metadata: {
											model: 'claude-sonnet-4-20250514',
											usage: {
												inputTokens: 100,
												outputTokens: 50,
												cacheCreationInputTokens: 20,
												cacheReadInputTokens: 10,
											},
										},
										credits: 1.25,
									},
								]),
							},
						},
					},
				},
			});
			vi.stubEnv('CODEBUFF_DATA_DIR', fixture.path);

			await expect(loadCodebuffUsageEntries()).resolves.toEqual([
				{
					timestamp: '2026-01-02T03:04:06.000Z',
					sessionId: `${path.basename(fixture.path)}/project-a/2026-01-02T03-04-05.000Z`,
					model: 'claude-sonnet-4-20250514',
					provider: 'anthropic',
					credits: 1.25,
					inputTokens: 100,
					outputTokens: 50,
					cacheCreationInputTokens: 20,
					cacheReadInputTokens: 10,
					dedupKey: `codebuff:${path.basename(fixture.path)}/project-a/2026-01-02T03-04-05.000Z:assistant-message`,
				},
			]);
		});

		it('keeps upstream message ids scoped to their chat sessions', async () => {
			await using fixture = await createFixture({
				projects: {
					project: {
						chats: {
							'2026-01-02T03-04-05.000Z': {
								'chat-messages.json': JSON.stringify([
									{
										id: 'assistant-message',
										role: 'assistant',
										metadata: {
											model: 'claude-sonnet-4-20250514',
											usage: { inputTokens: 100, outputTokens: 50 },
										},
									},
								]),
							},
							'2026-01-03T03-04-05.000Z': {
								'chat-messages.json': JSON.stringify([
									{
										id: 'assistant-message',
										role: 'assistant',
										metadata: {
											model: 'claude-sonnet-4-20250514',
											usage: { inputTokens: 200, outputTokens: 60 },
										},
									},
								]),
							},
						},
					},
				},
			});
			vi.stubEnv('CODEBUFF_DATA_DIR', fixture.path);

			const entries = await loadCodebuffUsageEntries();
			expect(entries).toHaveLength(2);
			expect(entries.map((entry) => entry.dedupKey)).toEqual([
				`codebuff:${path.basename(fixture.path)}/project/2026-01-02T03-04-05.000Z:assistant-message`,
				`codebuff:${path.basename(fixture.path)}/project/2026-01-03T03-04-05.000Z:assistant-message`,
			]);
		});

		it('classifies cachedTokensCreated as cache creation tokens', async () => {
			await using fixture = await createFixture({
				projects: {
					project: {
						chats: {
							'2026-01-02T03-04-05.000Z': {
								'chat-messages.json': JSON.stringify([
									{
										role: 'assistant',
										metadata: {
											model: 'claude-sonnet-4-20250514',
											usage: {
												inputTokens: 100,
												outputTokens: 50,
												cachedTokensCreated: 25,
											},
										},
									},
								]),
							},
						},
					},
				},
			});
			vi.stubEnv('CODEBUFF_DATA_DIR', fixture.path);

			await expect(loadCodebuffUsageEntries()).resolves.toMatchObject([
				{
					cacheCreationInputTokens: 25,
					cacheReadInputTokens: 0,
				},
			]);
		});

		it('falls back to runState provider usage and chat id timestamps', async () => {
			await using fixture = await createFixture({
				projects: {
					'project-a': {
						chats: {
							'2026-01-02T03-04-05.000Z': {
								'chat-messages.json': JSON.stringify([
									{
										variant: 'agent',
										metadata: {
											runState: {
												sessionState: {
													mainAgentState: {
														messageHistory: [
															{ role: 'user', providerOptions: {} },
															{
																role: 'assistant',
																providerOptions: {
																	codebuff: {
																		model: 'openai/gpt-5',
																		usage: {
																			prompt_tokens: 100,
																			completion_tokens: 50,
																			prompt_tokens_details: { cached_tokens: 10 },
																		},
																	},
																},
															},
														],
													},
												},
											},
										},
									},
								]),
							},
						},
					},
				},
			});
			vi.stubEnv('CODEBUFF_DATA_DIR', fixture.path);

			await expect(loadCodebuffUsageEntries()).resolves.toMatchObject([
				{
					timestamp: '2026-01-02T03:04:05.000Z',
					model: 'openai/gpt-5',
					provider: 'openai',
					inputTokens: 100,
					outputTokens: 50,
					cacheReadInputTokens: 10,
				},
			]);
		});
	});

	describe('parseCodebuffChatIdTimestamp', () => {
		it('restores time separators after the date part', () => {
			expect(parseCodebuffChatIdTimestamp('2026-01-02T03-04-05.000Z')).toBe(
				'2026-01-02T03:04:05.000Z',
			);
		});
	});
}

File: rust/crates/ccusage/src/adapter/codebuff/parser.rs
use std::{fs, path::Path};

use serde_json::Value;

use crate::{
    PricingMap, Result, TokenUsageRaw, apply_total_token_fallback, calculate_cost_for_usage,
    cli::CostMode, format_rfc3339_millis, missing_pricing_model_for_candidates, parse_ts_timestamp,
};

const DEFAULT_CODEBUFF_MODEL: &str = "codebuff-unknown";

#[derive(Clone, Default)]
pub(super) struct AssistantUsage {
    pub(super) model: Option<String>,
    pub(super) credits: f64,
    pub(super) input_tokens: u64,
    pub(super) output_tokens: u64,
    pub(super) cache_creation_input_tokens: u64,
    pub(super) cache_read_input_tokens: u64,
    pub(super) extra_total_tokens: u64,
}

pub(super) struct CodebuffEntry {
    pub(super) timestamp: crate::TimestampMs,
    pub(super) timestamp_text: String,
    pub(super) session_id: String,
    pub(super) model: String,
    provider: String,
    pub(super) credits: f64,
    pub(super) usage: TokenUsageRaw,
    pub(super) extra_total_tokens: u64,
    pub(super) dedup_key: String,
}

struct CodebuffContext {
    chat_id: String,
    session_id: String,
}

pub(super) fn load_chat_file(path: &Path) -> Result<Vec<CodebuffEntry>> {
    let content = fs::read_to_string(path)?;
    let Ok(messages) = serde_json::from_str::<Value>(&content) else {
        return Ok(Vec::new());
    };
    let Some(messages) = messages.as_array() else {
        return Ok(Vec::new());
    };
    let context = derive_context(path);
    let chat_timestamp = parse_codebuff_chat_id_timestamp(&context.chat_id);
    let file_timestamp = file_modified_timestamp(path).unwrap_or(crate::TimestampMs::UNIX_EPOCH);
    let mut entries = Vec::new();
    for (ordinal, message) in messages.iter().enumerate() {
        let Some(message) = message.as_object() else {
            continue;
        };
        if !is_assistant_message(message) {
            continue;
        }
        let usage = extract_assistant_usage(message);
        if !has_signal(&usage) {
            continue;
        }
        let timestamp = message_timestamp(message)
            .or(chat_timestamp)
            .unwrap_or(file_timestamp);
        let model = usage
            .model
            .clone()
            .unwrap_or_else(|| DEFAULT_CODEBUFF_MODEL.to_string());
        let dedup_key = dedup_key(
            message,
            &context.session_id,
            timestamp,
            &model,
            &usage,
            ordinal,
        );
        entries.push(CodebuffEntry {
            timestamp,
            timestamp_text: format_rfc3339_millis(timestamp),
            session_id: context.session_id.clone(),
            provider: infer_provider(&model).to_string(),
            model,
            credits: usage.credits,
            usage: TokenUsageRaw {
                input_tokens: usage.input_tokens,
                output_tokens: usage.output_tokens,
                cache_creation_input_tokens: usage.cache_creation_input_tokens,
                cache_read_input_tokens: usage.cache_read_input_tokens,
                speed: None,
                cache_creation: None,
            },
            extra_total_tokens: usage.extra_total_tokens,
            dedup_key,
        });
    }
    Ok(entries)
}

fn derive_context(path: &Path) -> CodebuffContext {
    let chat_id = path
        .parent()
        .and_then(Path::file_name)
        .and_then(|name| name.to_str())
        .filter(|name| !name.is_empty())
        .unwrap_or("unknown")
        .to_string();
    let chats_dir = path.parent().and_then(Path::parent);
    let project_dir = chats_dir.and_then(Path::parent);
    let project = project_dir
        .and_then(Path::file_name)
        .and_then(|name| name.to_str())
        .filter(|name| !name.is_empty())
        .unwrap_or("unknown")
        .to_string();
    let channel = project_dir
        .and_then(Path::parent)
        .and_then(Path::parent)
        .and_then(Path::file_name)
        .and_then(|name| name.to_str())
        .filter(|name| !name.is_empty())
        .unwrap_or("manicode")
        .to_string();
    CodebuffContext {
        session_id: format!("{channel}/{project}/{chat_id}"),
        chat_id,
    }
}

fn is_assistant_message(message: &serde_json::Map<String, Value>) -> bool {
    matches!(
        string_field(message, "variant")
            .or_else(|| string_field(message, "role"))
            .as_deref(),
        Some("ai" | "agent" | "assistant")
    )
}

fn extract_assistant_usage(message: &serde_json::Map<String, Value>) -> AssistantUsage {
    let mut usage = AssistantUsage::default();
    if let Some(metadata) = object_field(message, "metadata") {
        usage.model = string_field(metadata, "model");
        merge_fallback(&mut usage, parse_usage_object(metadata.get("usage")));
        merge_fallback(
            &mut usage,
            parse_usage_object(
                metadata
                    .get("codebuff")
                    .and_then(Value::as_object)
                    .and_then(|codebuff| codebuff.get("usage")),
            ),
        );
        if let Some(run_state_usage) = extract_usage_from_run_state(metadata) {
            merge_fallback(&mut usage, run_state_usage);
        }
    }
    let credits = number_field(message, "credits");
    if credits > 0.0 && usage.credits <= 0.0 {
        usage.credits = credits;
    }
    usage
}

fn extract_usage_from_run_state(
    metadata: &serde_json::Map<String, Value>,
) -> Option<AssistantUsage> {
    let history = metadata
        .get("runState")?
        .get("sessionState")?
        .get("mainAgentState")?
        .get("messageHistory")?
        .as_array()?;
    let mut usage = AssistantUsage::default();
    let mut found = false;
    for item in history.iter().rev() {
        let Some(entry) = item.as_object() else {
            continue;
        };
        if string_field(entry, "role").as_deref() != Some("assistant") {
            continue;
        }
        let Some(provider_options) = object_field(entry, "providerOptions") else {
            continue;
        };
        let mut entry_usage = AssistantUsage::default();
        merge_fallback(
            &mut entry_usage,
            parse_usage_object(provider_options.get("usage")),
        );
        if let Some(codebuff) = object_field(provider_options, "codebuff") {
            merge_fallback(&mut entry_usage, parse_usage_object(codebuff.get("usage")));
            entry_usage.model = string_field(codebuff, "model").or(entry_usage.model);
        }
        if has_signal(&entry_usage) || entry_usage.model.is_some() {
            found = true;
        }
        merge_fallback(&mut usage, entry_usage);
    }
    found.then_some(usage)
}

pub(super) fn parse_usage_object(value: Option<&Value>) -> AssistantUsage {
    let mut usage = AssistantUsage::default();
    let Some(record) = value.and_then(Value::as_object) else {
        return usage;
    };
    usage.input_tokens = pick_u64(
        record,
        &[
            "inputTokens",
            "input_tokens",
            "promptTokens",
            "prompt_tokens",
        ],
    );
    usage.output_tokens = pick_u64(
        record,
        &[
            "outputTokens",
            "output_tokens",
            "completionTokens",
            "completion_tokens",
        ],
    );
    usage.cache_read_input_tokens =
        pick_u64(record, &["cacheReadInputTokens", "cache_read_input_tokens"])
            .max(pick_nested_u64(
                record,
                "promptTokensDetails",
                &["cachedTokens"],
            ))
            .max(pick_nested_u64(
                record,
                "prompt_tokens_details",
                &["cached_tokens"],
            ));
    usage.cache_creation_input_tokens = pick_u64(
        record,
        &[
            "cacheCreationInputTokens",
            "cache_creation_input_tokens",
            "cacheCreationTokens",
            "cache_creation_tokens",
            "cachedTokensCreated",
            "cached_tokens_created",
        ],
    );
    let total_tokens = pick_u64(record, &["totalTokens", "total_tokens", "total"]);
    let raw_usage = TokenUsageRaw {
        input_tokens: usage.input_tokens,
        output_tokens: usage.output_tokens,
        cache_creation_input_tokens: usage.cache_creation_input_tokens,
        cache_read_input_tokens: usage.cache_read_input_tokens,
        speed: None,
        cache_creation: None,
    };
    let (raw_usage, extra_total_tokens) =
        apply_total_token_fallback(raw_usage, usage.extra_total_tokens, total_tokens);
    usage.input_tokens = raw_usage.input_tokens;
    usage.output_tokens = raw_usage.output_tokens;
    usage.cache_creation_input_tokens = raw_usage.cache_creation_input_tokens;
    usage.cache_read_input_tokens = raw_usage.cache_read_input_tokens;
    usage.extra_total_tokens = extra_total_tokens;
    usage.credits = number_field(record, "credits");
    usage.model = string_field(record, "model");
    usage
}

fn merge_fallback(target: &mut AssistantUsage, fallback: AssistantUsage) {
    if target.input_tokens == 0 {
        target.input_tokens = fallback.input_tokens;
    }
    if target.output_tokens == 0 {
        target.output_tokens = fallback.output_tokens;
    }
    if target.cache_creation_input_tokens == 0 {
        target.cache_creation_input_tokens = fallback.cache_creation_input_tokens;
    }
    if target.cache_read_input_tokens == 0 {
        target.cache_read_input_tokens = fallback.cache_read_input_tokens;
    }
    if target.extra_total_tokens == 0 {
        target.extra_total_tokens = fallback.extra_total_tokens;
    }
    if target.credits <= 0.0 {
        target.credits = fallback.credits;
    }
    if target.model.is_none() {
        target.model = fallback.model;
    }
}

fn has_signal(usage: &AssistantUsage) -> bool {
    usage.input_tokens > 0
        || usage.output_tokens > 0
        || usage.cache_creation_input_tokens > 0
        || usage.cache_read_input_tokens > 0
        || usage.extra_total_tokens > 0
        || usage.credits > 0.0
}

fn message_timestamp(message: &serde_json::Map<String, Value>) -> Option<crate::TimestampMs> {
    timestamp_value(message.get("timestamp"))
        .or_else(|| timestamp_value(message.get("createdAt")))
        .or_else(|| {
            object_field(message, "metadata")
                .and_then(|metadata| timestamp_value(metadata.get("timestamp")))
        })
}

fn parse_codebuff_chat_id_timestamp(chat_id: &str) -> Option<crate::TimestampMs> {
    let (date, time) = chat_id.split_once('T')?;
    let mut time = time.to_string();
    for _ in 0..2 {
        if let Some(index) = time.find('-') {
            time.replace_range(index..=index, ":");
        }
    }
    parse_ts_timestamp(&format!("{date}T{time}"))
}

fn timestamp_value(value: Option<&Value>) -> Option<crate::TimestampMs> {
    match value? {
        Value::String(value) => parse_ts_timestamp(value),
        Value::Number(value) => {
            let raw = value.as_i64()?;
            let millis = if raw < 10_000_000_000 {
                raw.checked_mul(1_000)?
            } else {
                raw
            };
            (millis > 0).then(|| crate::TimestampMs::from_millis(millis))
        }
        _ => None,
    }
}

fn file_modified_timestamp(path: &Path) -> Option<crate::TimestampMs> {
    let modified = fs::metadata(path).ok()?.modified().ok()?;
    let millis = modified
        .duration_since(std::time::UNIX_EPOCH)
        .ok()?
        .as_millis()
        .min(i64::MAX as u128) as i64;
    Some(crate::TimestampMs::from_millis(millis))
}

fn dedup_key(
    message: &serde_json::Map<String, Value>,
    session_id: &str,
    timestamp: crate::TimestampMs,
    model: &str,
    usage: &AssistantUsage,
    ordinal: usize,
) -> String {
    if let Some(id) = string_field(message, "id") {
        return format!("codebuff:{session_id}:{id}");
    }
    format!(
        "codebuff:{session_id}:{}:{model}:{ordinal}:{}:{}:{}:{}:{}",
        format_rfc3339_millis(timestamp),
        usage.input_tokens,
        usage.output_tokens,
        usage.cache_read_input_tokens,
        usage.cache_creation_input_tokens,
        usage.extra_total_tokens
    )
}

fn infer_provider(model: &str) -> &'static str {
    let model = model.to_ascii_lowercase();
    if model.starts_with("claude-")
        || model.starts_with("anthropic/")
        || model.starts_with("anthropic.")
    {
        "anthropic"
    } else if model.starts_with("gpt-")
        || model.starts_with("o1")
        || model.starts_with("o3")
        || model.starts_with("o4")
        || model.starts_with("openai/")
    {
        "openai"
    } else if model.starts_with("gemini") || model.starts_with("google/") {
        "google"
    } else if model.starts_with("grok") || model.starts_with("xai/") {
        "xai"
    } else if model.starts_with("openrouter/") {
        "openrouter"
    } else {
        "unknown"
    }
}

pub(super) fn calculate_codebuff_cost(entry: &CodebuffEntry, pricing: &PricingMap) -> f64 {
    let usage = TokenUsageRaw {
        output_tokens: entry
            .usage
            .output_tokens
            .saturating_add(entry.extra_total_tokens),
        cache_creation: None,
        ..entry.usage
    };
    let raw = calculate_cost_for_usage(
        Some(&entry.model),
        usage,
        None,
        CostMode::Calculate,
        Some(pricing),
    );
    if raw > 0.0
        || entry.provider == "unknown"
        || entry.model.starts_with(&format!("{}/", entry.provider))
    {
        return raw;
    }
    calculate_cost_for_usage(
        Some(&format!("{}/{}", entry.provider, entry.model)),
        usage,
        None,
        CostMode::Calculate,
        Some(pricing),
    )
}

pub(super) fn missing_codebuff_pricing(
    entry: &CodebuffEntry,
    pricing: &PricingMap,
) -> Option<String> {
    let usage = TokenUsageRaw {
        output_tokens: entry
            .usage
            .output_tokens
            .saturating_add(entry.extra_total_tokens),
        cache_creation: None,
        ..entry.usage
    };
    let mut candidates = vec![entry.model.clone()];
    if entry.provider != "unknown" && !entry.model.starts_with(&format!("{}/", entry.provider)) {
        candidates.push(format!("{}/{}", entry.provider, entry.model));
    }
    missing_pricing_model_for_candidates(
        &entry.model,
        candidates,
        crate::total_usage_tokens(usage),
        Some(pricing),
    )
}

fn object_field<'a>(
    record: &'a serde_json::Map<String, Value>,
    key: &str,
) -> Option<&'a serde_json::Map<String, Value>> {
    record.get(key).and_then(Value::as_object)
}

fn string_field(record: &serde_json::Map<String, Value>, key: &str) -> Option<String> {
    let value = record.get(key)?.as_str()?.trim();
    (!value.is_empty()).then(|| value.to_string())
}

fn number_field(record: &serde_json::Map<String, Value>, key: &str) -> f64 {
    record
        .get(key)
        .and_then(Value::as_f64)
        .filter(|value| value.is_finite() && *value > 0.0)
        .unwrap_or(0.0)
}

fn pick_u64(record: &serde_json::Map<String, Value>, keys: &[&str]) -> u64 {
    keys.iter()
        .filter_map(|key| record.get(*key))
        .find_map(|value| value.as_u64().filter(|value| *value > 0))
        .unwrap_or(0)
}

fn pick_nested_u64(record: &serde_json::Map<String, Value>, key: &str, keys: &[&str]) -> u64 {
    object_field(record, key).map_or(0, |nested| pick_u64(nested, keys))
}
